testsuite/shared/src/main/java/org/jboss/as/test/shared/observability/containers/OpenTelemetryCollectorContainer.java

```diff
@@ -76,26 +89,76 @@ public List<JaegerTrace> getTraces(String serviceName) throws InterruptedExcepti
         return (jaegerContainer != null ? jaegerContainer.getTraces(serviceName) : Collections.emptyList());
     }
 
-    public List<PrometheusMetric> fetchMetrics(String nameToMonitor) throws InterruptedException {
-        String body = "";
-        try (Client client = ClientBuilder.newClient()) {
-            WebTarget target = client.target(getPrometheusUrl());
+    Duration DEFAULT_TIMEOUT = Duration.ofSeconds(TimeoutUtil.adjust(30));
+
+    /**
+     * Continually evaluates assertions provided in a consumer until the state obtained from the Prometheus endpoint
+     * matches the expected state or until a timeout elapses.
+     * By default, polls the collector every second for 30 seconds.
+     * Returns snapshot of the prometheus registry that passed the assertions; typically ignored.
+     *
+     * @param assertionConsumer consumer implementation that contains {@link Assert}ions throwing {@link AssertionError#AssertionError()}s
+     *                          if the state obtained from the Prometheus endpoint does not match the expected state
+     * @return list of prometheus metrics; typically ignored.
+     * @throws AssertionError if
+     * @throws InterruptedException if interrupted
+     */
+    public List<PrometheusMetric> assertMetrics(Consumer<List<PrometheusMetric>> assertionConsumer) throws AssertionError, InterruptedException {
+        return this.assertMetrics(assertionConsumer, DEFAULT_TIMEOUT);
+    }
 
-            int attemptCount = 0;
-            boolean found = false;
+    /**
+     * Variant of {@link OpenTelemetryCollectorContainer#assertMetrics(Consumer)} that can be configured with a timeout duration.
+     */
+    public List<PrometheusMetric> assertMetrics(Consumer<List<PrometheusMetric>> assertionConsumer, Duration timeout) throws AssertionError, InterruptedException {
+        debugLog("assertMetrics(..) validation starting.");
 
-            // Request counts can vary. Setting high to help ensure test stability
-            while (!found && attemptCount < 30) {
-                // Wait to give Micrometer time to export
-                Thread.sleep(1000);
+        Instant endTime = Instant.now().plus(timeout);
+
+        AssertionError lastAssertionError = null;
+
+        while (Instant.now().isBefore(endTime)) {
+            try {
+                List<PrometheusMetric> prometheusMetrics = fetchMetrics();
+
+                assertionConsumer.accept(prometheusMetrics);
 
-                body = target.request().get().readEntity(String.class);
-                found = body.contains(nameToMonitor);
-                attemptCount++;
+                debugLog("assertMetrics(..) validation passed.");
+
+                return prometheusMetrics;
+            } catch (AssertionError assertionError) {
+                debugLog("assertMetrics(..) validation failed - retrying.");
+                lastAssertionError = assertionError;
+                Thread.sleep(1000);
```

Comment:
ThreadUtils.sleep(Duration.ofMillis(1000));

Reply:
Hrm. I think I was thinking of TimeoutUtils, which you're already using. Take or leave this one. :)